pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,489 @@
+	/// on the routers that we have in storage.
+	pub fn get_router_ids_for_domain(domain: Domain) -> Result<Vec<T::RouterId>, DispatchError> {
+		let all_routers_for_domain = T::RouterProvider::routers_for_domain(domain);
+
+		let stored_routers = Routers::<T>::get();
+
+		let res = all_routers_for_domain
+			.iter()
+			.filter(|x| stored_routers.iter().any(|y| *x == y))
+			.map(|x| x.clone())
+			.collect::<Vec<_>>();
+
+		if res.is_empty() {
+			return Err(Error::<T>::NotEnoughRoutersForDomain.into());
+		}
+
+		Ok(res)
+	}
+
+	/// Calculates and returns the proof count required for processing one
+	/// inbound message.
+	fn get_expected_proof_count(router_ids: &Vec<T::RouterId>) -> Result<u32, DispatchError> {
+		let expected_proof_count = router_ids
+			.len()
+			.ensure_sub(1)
+			.map_err(|_| Error::<T>::NotEnoughRoutersForDomain)?;
+
+		Ok(expected_proof_count as u32)
+	}
+
+	/// Gets the message proof for a message.
+	fn get_message_proof(message: T::Message) -> Proof {
+		match message.get_message_proof() {
+			None => message
+				.to_message_proof()
+				.get_message_proof()
+				.expect("message proof ensured by 'to_message_proof'"),
+			Some(proof) => proof,
+		}
+	}
+
+	/// Creates an inbound entry based on whether the inbound message is a
+	/// proof or not.
+	fn create_inbound_entry(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		message: T::Message,
+	) -> InboundEntry<T> {
+		match message.get_message_proof() {
+			None => InboundEntry::Message {
+				session_id: inbound_processing_info.current_session_id.clone(),
+				domain_address: inbound_processing_info.domain_address.clone(),
+				message,
+				expected_proof_count: inbound_processing_info.expected_proof_count_per_message,
+			},
+			Some(_) => InboundEntry::Proof {
+				session_id: inbound_processing_info.current_session_id,
+				current_count: 1,
+			},
+		}
+	}
+
+	/// Validation ensures that:
+	///
+	/// - the router that sent the inbound message is a valid router for the
+	///   specific domain.
+	/// - messages are only sent by the first inbound router.
+	/// - proofs are not sent by the first inbound router.
+	fn validate_inbound_entry(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		router_id: &T::RouterId,
+		inbound_entry: &InboundEntry<T>,
+	) -> DispatchResult {
+		let router_ids = inbound_processing_info.router_ids.clone();
+
+		ensure!(
+			router_ids.iter().any(|x| x == router_id),
+			Error::<T>::UnknownRouter
+		);
+
+		match inbound_entry {
+			InboundEntry::Message { .. } => {
+				ensure!(
+					router_ids.get(0) == Some(&router_id),
+					Error::<T>::MessageExpectedFromFirstRouter
+				);
+
+				Ok(())
+			}
+			InboundEntry::Proof { .. } => {
+				ensure!(
+					router_ids.get(0) != Some(&router_id),
+					Error::<T>::ProofNotExpectedFromFirstRouter
+				);
+
+				Ok(())
+			}
+		}
+	}
+
+	/// Upserts an inbound entry for a particular message, increasing the
+	/// relevant counts accordingly.
+	fn upsert_pending_entry(
```

Comment:
Nit: I have to admit it's really difficult to follow this function. If we have the time on Friday, let's improve readability by splitting into smaller chunks if possible, e.g. impl compare functions for `InboundEntry`.

Reply:
I agree, I've seen it too much lately so it's a bit easier for me but still... I'll take a stab at this soon and try to break it into smaller pieces.

Reply:
I think nested matches generally make things difficult to read. Maybe 
```rust
match (stored_entry, entry)
// four cases
```
works better? Not sure.

Also the struct-enum like forces you to always match in the fields, which becomes a bit noisy. Maybe splitting these enums into an enum with `MessageEntry` and `MessageProof` structures makes things much simpler:

```rust
enum InboundEntry {
   Message(MessageEntry)
   Proof(ProofEntry)
}
```

Reply:
I already moved the update logic to a `fn update(&mut self, other: Self)` for inbound entries. Will try out your suggestions there and hopefully we will have something that doesn't cause migraines on review.

Reply:
Done, PTAL <3